many fixes to emit working declarations

boneskull opened the pull request:
This fixes the invalid declarations generated by TypeScript by rewriting `@appium/base-driver`'s
`lib/basedriver/driver.js` (containing the `BaseDriver` class) in TypeScript.

It also addresses a host of issues found when doing so, particularly around the use of the
`*Capabilities` and `*DriverCaps` types.  The question of which of these an external driver should
use is much less murky (e.g., you want to use `createSession(caps: W3CDriverCaps<MyConstraints>)`).
See the detailed commit message for more information.

It builds upon the structure added in #18390 by adding handful of simple checks for `BaseDriver`.

Closes #18379

boneskull committed:
chore(base-driver): add a failing type test

boneskull committed:
fix(appium,types,base-driver,fake-driver,driver-test-support): normalize constraint defaults

Another handful of fixes.

- remove second "extra" parameter from `Capabilities`, `NSCapabilities` `W3CCapabilities`. we were not using it anywhere to my knowledge. it was an escape hatch anyway.
- never assign `BaseDriverCapConstraints` as the default constraints, because this is confusing as hell. instead, drivers should use the `*DriverCaps` types which automatically pulls the base constraints in. applied across several workspaces
- renamed `DeviceSettings` => `IDeviceSettings` because `DeviceSettings` is a class.
- the session ID param of `Core`'s `sessionExists()` is actually optional (per the implementation)
- allow some modicum of type narrowing at the class level around settings and values returned by `ISessionHandler`'s methods. also, not all values in the raw settings object are of the same type `T`, so fix that
- `server`, `serverHost`, `serverPort` and `serverPath` are not members of `Core` (using `AppiumDriver` as a litmus test--it interacts with `Driver` types but does not have these properties itself)
- changed a few types to interfaces where possible
- moved a lost comment
- further dumbed-down the logdef stuff
- `Driver.driverData` is _always_ defined, since `Core` makes it an empty array.
- `DriverOpts` no longer contains "partial" caps; this it not quite right as the caps are not assigned until a session is created, but there's a conflict between some props which are _both_ server args and defined in constraints.  anyway, this is what `initialOpts` is for
- fix some type disagreements in basedriver's "capabilities" module

boneskull committed:
fix(basedriver): convert main driver implementation to typescript

This seems to fix the bug in #18379.  There must be some sort of bug in TypeScript w/r/t `@implements` or maybe even it was not treating `@typedef` as a proper alias.  Perhaps it's user error--omitting optional type args in alias-style typedefs may end up making a new type entirely.  idk, but this was easier to deal with.

Closes #18379.

boneskull commented:
Current dependencies on/for this PR:
* master
  * **PR #18390** <a href="https://app.graphite.dev/github/pr/appium/appium/18390" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #18391** <a href="https://app.graphite.dev/github/pr/appium/appium/18391" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/appium/appium/18391?utm_source=stack-comment).

boneskull committed:
fix: backwards-compatible fixes for TS v5.x

These will be issues when we upgrade

jlipps commented on packages/base-driver/lib/protocol/routes.js:
```diff
@@ -18,6 +18,7 @@ const SET_ALERT_TEXT_PAYLOAD_PARAMS = {
  * define the routes, mapping of HTTP methods to particular driver commands, and
  * any parameters that are expected in a request parameters can be `required` or
  * `optional`
+ * @satisfies {import('@appium/types').MethodMap<import('../basedriver/driver').BaseDriver>}
```
oh cool, is this new?

boneskull replied:
yes!

boneskull committed:
fix(types): fix method map types
These types do not care about constraints, only the shape of the driver itself.
